crates/autopilot/src/decoded_settlement.rs

```diff
@@ -0,0 +1,511 @@
+                        executed_amount,
+                        signature,
+                    },
+                )
+                .collect(),
+            interactions: [
+                interactions[0]
+                    .clone()
+                    .into_iter()
+                    .map(Into::into)
+                    .collect(),
+                interactions[1]
+                    .clone()
+                    .into_iter()
+                    .map(Into::into)
+                    .collect(),
+                interactions[2]
+                    .clone()
+                    .into_iter()
+                    .map(Into::into)
+                    .collect(),
+            ],
+        }
+    }
+}
+
+pub struct FeeConfiguration {
+    pub fee_objective_scaling_factor: f64,
+}
+
+impl DecodedSettlement {
+    pub fn new(contract: &GPv2Settlement, input: &[u8]) -> Result<Self> {
+        let function = contract
+            .raw_instance()
+            .abi()
+            .function("settle")
+            .context("settle function not found")?;
+        let decoded_input = decode_function_input(function, input)?;
+        <DecodedSettlementTokenized>::from_token(Token::Tuple(decoded_input))
+            .map(Into::into)
+            .context("failed to decode settlement")
+    }
+
+    /// Returns the total surplus denominated in the native asset for the
+    /// solution.
+    pub fn total_surplus(&self, external_prices: &ExternalPrices) -> U256 {
+        self.trades.iter().fold(0.into(), |acc, trade| {
+            acc + surplus(trade, &self.tokens, &self.clearing_prices, external_prices)
+                .unwrap_or_default()
+        })
+    }
+
+    // Assumes it is called with already FILLED orders.
+    // Needs rework to support partially fillable orders.
+    // Tricky because the decoded settlement is using FILLED `orders` so we don't
+    // always know the executed amount in case of partial fill.
+    pub fn total_fees(
+        &self,
+        external_prices: &ExternalPrices,
+        orders: &[Order],
+        configuration: &FeeConfiguration,
+    ) -> U256 {
+        self.trades.iter().fold(0.into(), |acc, trade| {
+            match orders.iter().find(|order| {
+                let signature = Bytes(
+                    order
+                        .signature
+                        .encode_for_settlement(order.metadata.owner)
+                        .to_vec(),
+                );
+                signature == trade.signature
+            }) {
+                Some(order) => {
+                    acc + fee(trade, external_prices, order, configuration).unwrap_or_default()
+                }
+                None => {
+                    tracing::warn!(
+                        "order not found for trade, possible incomplete fee calculation"
+                    );
+                    acc
+                }
+            }
+        })
+    }
+}
+
+fn surplus(
+    trade: &DecodedTrade,
```

Comment:
Surplus calculation is massively copy-pasted from `Settlement` struct. Code duplication could be removed if we merge `DecodedSettlement` and `EncodedSettlement` into one.